src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSyncController.scala

```diff
@@ -0,0 +1,195 @@
+package io.iohk.ethereum.blockchain.sync
+
+import akka.actor._
+import io.iohk.ethereum.blockchain.sync.SyncController.StartRegularSync
+import io.iohk.ethereum.domain.BlockHeader
+import io.iohk.ethereum.network.PeerActor.Status.{Chain, Handshaked}
+import io.iohk.ethereum.network.PeerActor._
+import io.iohk.ethereum.network.p2p.messages.CommonMessages.NewBlock
+import io.iohk.ethereum.network.p2p.messages.PV62._
+import io.iohk.ethereum.utils.Config
+import org.spongycastle.util.encoders.Hex
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.util.Try
+
+trait RegularSyncController {
+  self: SyncController =>
+
+  private var headersQueue: Seq[BlockHeader] = Seq.empty
+  private var resolvingBranch = false
+  private var broadcasting = false
+  private var timeout: Option[Cancellable] = None
+
+  import Config.FastSync._
+
+  def regularSync(): Receive = {
+
+    case StartRegularSync =>
+      askForHeaders()
+
+    case ResumeRegularSync =>
+      askForHeaders()
+
+    case MessageReceived(m: BlockHeaders) if !resolvingBranch =>
+      cancelTimeout()
+      handleDownload(m)
+
+    case MessageReceived(m: BlockHeaders) if resolvingBranch =>
+      cancelTimeout()
+      handleBlockBranchResolution(m)
+
+    case MessageReceived(m: BlockBodies) =>
+      cancelTimeout()
+      handleBlockBodies(m)
+
+    case m: BroadcastBlocks if broadcasting =>
+      peersToDownloadFrom.keys.foreach(_ ! m)
+
+    case ResolveBranch(peer) =>
+      requestHeadersForNewBranch(peer)
+
+    case PeerTimeOut(peer) =>
+      blacklist(peer, blacklistDuration)
+      resolvingBranch = false
+      askForHeaders()
+  }
+
+  private def requestHeadersForNewBranch(peer: ActorRef) = {
+    resolvingBranch = true
+    sendWithTimeout(peer, SendMessage(GetBlockHeaders(Right(headersQueue.head.parentHash), blockResolveDepth, skip = 0, reverse = true)))
+  }
+
+  private def askForHeaders() = {
+
+    bestPeer match {
+      case Some(peer) =>
+        val blockNumber = appStateStorage.getBestBlockNumber()
+
+        peer ! Subscribe(Set(BlockHeaders.code, BlockBodies.code))
+        sendWithTimeout(peer, SendMessage(GetBlockHeaders(Left(blockNumber + 1), blockHeadersPerRequest, skip = 0, reverse = false)))
+      case None =>
+        log.warning("no peers to download from")
+        scheduleResume()
+    }
+  }
+
+  private def handleBlockBranchResolution(message: BlockHeaders) =
+    //todo limit max branch depth?
```

Comment:
Can we blacklist a peer which would require to drop too many blocks from currently processed branch?

I'd do that in a separate PR (and YT ticket).